### DISCUSSION TITLE
RFI on the implementation of F´ to ARINC653

### DISCUSSION THREAD

**Discussion by ThibFrgsGmz** [nasa/fprime#972] [Q&A] [answered]
Hi there 👋,

While reading the F Prime documentation, I saw that the F' has been ported to ARINC 653, see [here](https://github.com/nasa/fprime/blob/e165c914b0d0240eba8164c52f8856ce90bee3dc/docs/UsersGuide/user/full-intro.md#conclusion) in the Conclusion section.

Could you tell me more about this? Do you have an example of an open source implementation? 👀

Our prime contractors, at least in Europe, are asking us more and more to adopt a TSP (Time & Space Partitioning) architecture on our products.
So we have to deliver a library to drive our instruments in the form of a partition whose interfaces shall be compliant with the ARINC653 standard.

Having never implemented them, I am not really familiar with TSP and this avionics standard.

I hear some technical referents in my department talking quickly about the ARINC653 standard and from what I understand, it mainly defines standardized interfaces for partitions and how to schedule partitions.
So I'm having a hard time understanding porting F' to ARINC653.

Having some information could help me a lot to defend the use of F' to my technical referents to instantiate it on future projects in my company.

An example of F Prime implementation in a TSP architecture with ARINC653 standard would also be gold to convince them. 😃

**Comment by timcanham** [comment, marked as answer]
@ThibFrgsGmz 

We had a research task early in F Prime development where we used an ARINC simulation to move components between partitions. The partitions communicated with each other via the [hub](https://nasa.github.io/fprime/Architecture/FPrimeArchitectureShort.pdf) pattern (see the architecture slides) using message channels. You simply connect components to either end of the hub to communicate with each other. They don't know they are on separate partitions.

The porting process would essentially to build a deployment in F Prime for each partition, and then interconnect them using the hubs. The particulars of compilers, etc, is dependent on the ARINC platform, although my recollection is that they may have a common API.

**Reply by ThibFrgsGmz** [reply]
OK, thank you very much @timcanham ! Theoretically this seems to be what I was imagining, where the hubs would use the IPC (Inter-Partition Communication) provided by the hypervisor  ([XTratum](https://fentiss.com/products/hypervisor/) for example) to communicate.

Could you tell me your or your department's views on the TSP architecture? Is it being implemented more and more in projects?
Is it more complicated to implement than a more traditional architecture and that is why it was not implemented in the F Prime example projects?
